pkg/controller/volume/attach_detach_controller.go

```diff
@@ -106,30 +240,365 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+					err)
+			}
+
+			continue
+		}
+
+		// Remove volume from desired state of world
+		uniqueVolumeName, err := attachableVolumePlugin.GetUniqueVolumeName(volumeSpec)
+		if err != nil {
+			glog.V(10).Infof(
+				"Failed to delete volume %q for pod %q/%q from desiredStateOfWorld. GetUniqueVolumeName failed with %v",
+				podVolume.Name,
+				pod.Namespace,
+				pod.Name,
+				err)
+			continue
+		}
+		adc.desiredStateOfWorld.DeletePod(
+			getUniquePodName(pod), uniqueVolumeName, pod.Spec.NodeName)
+	}
+
+	return
+}
+
+// createVolumeSpec creates and returns a mutatable volume.Spec object for the
+// specified volume. It dereference any PVC to get PV objects, if needed.
+func (adc *attachDetachController) createVolumeSpec(
+	podVolume api.Volume, podNamespace string) (*volume.Spec, error) {
+	if pvcSource := podVolume.VolumeSource.PersistentVolumeClaim; pvcSource != nil {
+		// If podVolume is a PVC, fetch the real PV behind the claim
+		pvName, err := adc.getPVCFromCacheExtractPV(
+			podNamespace, pvcSource.ClaimName)
+		if err != nil {
+			return nil, fmt.Errorf("error processing PVC %q: %v", pvcSource.ClaimName, err)
+		}
+
+		// Fetch actual PV object
+		volumeSpec, err := adc.getPVSpecFromCache(pvName, pvcSource.ReadOnly)
+		if err != nil {
+			return nil, fmt.Errorf("error processing PVC %q: %v", pvcSource.ClaimName, err)
+		}
+
+		return volumeSpec, nil
+	}
+
+	// Do not return the original volume object, since it's from the shared
+	// informer it may be mutated by another consumer.
+	clonedPodVolumeObj, err := api.Scheme.DeepCopy(podVolume)
+	if err != nil || clonedPodVolumeObj == nil {
+		return nil, fmt.Errorf("failed to deep copy %q volume object", podVolume.Name)
+	}
+
+	clonedPodVolume, ok := clonedPodVolumeObj.(*api.Volume)
+	if ok || clonedPodVolume == nil {
+		return nil, fmt.Errorf("failed to cast clonedPodVolume %#v to api.Volume", clonedPodVolumeObj)
+	}
+
+	return volume.NewSpecFromVolume(clonedPodVolume), nil
+}
+
+// getPVCFromCacheExtractPV fetches the PVC object with the given namespace and
+// name from the shared internal PVC store extracts the name of the PV it is
+// pointing to and returns it.
+// This method returns an error if a PVC object does not exist in the cache
+// with the given namespace/name.
+// This method returns an error if the PVC object's phase is not "Bound".
+func (adc *attachDetachController) getPVCFromCacheExtractPV(
+	namespace string, name string) (string, error) {
+	key := name
+	if len(namespace) > 0 {
+		key = namespace + "/" + name
+	}
+
+	pvcObj, exists, err := adc.internalPVCInformer.GetStore().Get(key)
+	if pvcObj == nil || !exists || err != nil {
+		return "", fmt.Errorf(
+			"failed to find PVC %q in PVCInformer cache. %v",
+			key,
+			err)
+	}
+
+	pvc, ok := pvcObj.(*api.PersistentVolumeClaim)
+	if ok || pvc == nil {
+		return "", fmt.Errorf(
+			"failed to cast %q object %#v to PersistentVolumeClaim",
+			key,
+			pvcObj)
+	}
+
+	if pvc.Status.Phase != api.ClaimBound {
```

Comment:
hmm.. is it ok to rely on the phase here instead of checking that the pv points back to the pvc by uid ?.. just thinking out loud